ecsclient/ecsclient_test.go

```diff
@@ -0,0 +1,689 @@
+package ecsclient
+
+import (
+	"errors"
+	"reflect"
+	"sort"
+	"testing"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/ecs"
+	"github.com/golang/mock/gomock"
+	"github.com/jpillora/backoff"
+	"github.com/quintilesims/go-ecs-cleaner/mocks"
+)
+
+func setup(t *testing.T) (*gomock.Controller, *ECSClient, *mocks.MockECSAPI) {
+	ctrl := gomock.NewController(t)
+
+	e := NewECSClient()
+	e.Flags.Quiet = true
+
+	svc := mocks.NewMockECSAPI(ctrl)
+	e.Svc = svc
+
+	return ctrl, e, svc
+}
+
+func Test_CollectClusters(t *testing.T) {
+	ctrl, e, svc := setup(t)
+	defer ctrl.Finish()
+
+	expected := []string{"arn0", "arn1", "arn2"}
+
+	svc.EXPECT().
+		ListClusters(&ecs.ListClustersInput{
+			NextToken: nil,
+		}).
+		Return(&ecs.ListClustersOutput{
+			ClusterArns: []*string{aws.String("arn0"), aws.String("arn1")},
+			NextToken:   aws.String("a"),
+		}, nil)
+
+	svc.EXPECT().
+		ListClusters(&ecs.ListClustersInput{
+			NextToken: aws.String("a"),
+		}).
+		Return(&ecs.ListClustersOutput{
+			ClusterArns: []*string{aws.String("arn2")},
+			NextToken:   nil,
+		}, nil)
+
+	result, err := e.CollectClusters()
+	if err != nil {
+		t.Error(err)
+	}
+
+	if equal := reflect.DeepEqual(expected, result); !equal {
+		t.Errorf("Expected %v, got %v\n", expected, result)
+	}
+}
+
+func Test_CollectServices(t *testing.T) {
+	ctrl, e, svc := setup(t)
+	defer ctrl.Finish()
+
+	clusterARNs := []string{"cluster1", "cluster0"}
+
+	expected := map[string][]string{
+		"cluster0": []string{"service0", "service1"},
+		"cluster1": []string{"service2"},
+	}
+
+	// paginated result
+	svc.EXPECT().
+		ListServices(&ecs.ListServicesInput{
+			Cluster:   aws.String("cluster0"),
+			NextToken: nil,
+		}).
+		Return(&ecs.ListServicesOutput{
+			ServiceArns: []*string{aws.String("service0")},
+			NextToken:   aws.String("a"),
+		}, nil)
+
+	svc.EXPECT().
+		ListServices(&ecs.ListServicesInput{
+			Cluster:   aws.String("cluster0"),
+			NextToken: aws.String("a"),
+		}).
+		Return(&ecs.ListServicesOutput{
+			ServiceArns: []*string{aws.String("service1")},
+			NextToken:   nil,
+		}, nil)
+
+	// unpaginated result
+	svc.EXPECT().
+		ListServices(&ecs.ListServicesInput{
+			Cluster:   aws.String("cluster1"),
+			NextToken: nil,
+		}).
+		Return(&ecs.ListServicesOutput{
+			ServiceArns: []*string{aws.String("service2")},
+			NextToken:   nil,
+		}, nil)
+
+	result, err := e.CollectServices(clusterARNs)
+	if err != nil {
+		t.Error(err)
+	}
+
+	if equal := reflect.DeepEqual(expected, result); !equal {
```

Comment:
I was frequently told not to use reflection in tests in the past, so my knee-jerk reaction here was to say 'no'.... that being said, from the `DeepEqual` [docs](https://golang.org/pkg/reflect/#DeepEqual):

> Map values are deeply equal when all of the following are true: they are both nil or both non-nil, they have the same length, and either they are the same map object or their corresponding keys (matched using Go equality) map to deeply equal values.

This seems to be a lot more convenient than iterating through the maps and comparing values manually. Trying to see a downside to this but I can't rn.

Reply:
Yeah, it seems pretty alright to me so far, so I dunno

Reply:
Yeah, it seems pretty alright to me, soooooo /shrug